src/test/webServer.ts

```diff
@@ -0,0 +1,140 @@
+      await webServer._waitForAvailability(port);
+      process.env.PLAYWRIGHT_TEST_BASE_URL = `http://localhost:${port}`;
+      return webServer;
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+  }
+
+  private async _verifyFreePort(port: number) {
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<number> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    const { launchedProcess, kill } = await launchProcess({
+      command: this.config.command,
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this.config.env,
+      },
+      cwd: this.config.cwd,
+      stdio: 'pipe',
+      shell: true,
+    });
+    this._killProcess = kill;
+
+    launchedProcess.stdout.pipe(newProcessLogPrefixer()).pipe(process.stdout);
+    launchedProcess.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
+    if (!this.config.port)
+      launchedProcess.stdout.on('data', collectPort);
+    let processExitedWithNonZeroStatusCodeCallback = (error: Error) => { };
+    this._processExitedWithNonZeroStatusCode = new Promise((_, reject) => processExitedWithNonZeroStatusCodeCallback = reject);
+    launchedProcess.on('exit', code => processExitedWithNonZeroStatusCodeCallback(new Error(`WebServer was not able to start. Exit code: ${code}`)));
+    if (this.config.port)
+      return this.config.port;
+    console.log(`Starting WebServer port detection.`);
+    const detectedPort = await Promise.race([
+      this._processExitedWithNonZeroStatusCode,
+      collectPortPromise,
+    ]);
+    console.log(`Port ${detectedPort} by process '${this.config.command}' was automatically detected.`);
+    return detectedPort;
+  }
+
+  private async _waitForAvailability(port: number) {
+    const launchTimeout = this.config.timeout || 60 * 1000;
+    const cancellationToken = { canceled: false };
+    const { timedOut } = (await Promise.race([
+      raceAgainstDeadline(waitForSocket(port, 100, cancellationToken), launchTimeout + monotonicTime()),
+      this._processExitedWithNonZeroStatusCode,
+    ]));
+    cancellationToken.canceled = true;
+    if (timedOut)
+      throw new Error(`failed to start web server on port ${port} via "${this.config.command}"`);
+  }
+  public async kill() {
+    await this._killProcess?.();
```

Comment:
This is not async anymore, is it?

Reply:
Good catch, types were wrong of the processLauncher, fixed it.